tempesta_fw/http_parser.c

```diff
@@ -2251,6 +2312,116 @@ __req_parse_host(TfwHttpReq *req, unsigned char *data, size_t len)
 	return r;
 }
 
+/**
+ * Function for parsing 'Referer' header value. This is a partial
+ * copy from parsing procedure for request start line URI with some 
+ * simplifications according to RFC 7231 chapter 5.5.2.
+ */
+static int
+__req_parse_referer(TfwHttpMsg *hm, unsigned char *data, size_t len)
+{
+	int r = CSTR_NEQ;
+	unsigned char *off_ptr = data;
+	__FSM_DECLARE_VARS(hm);
+
+	__FSM_START(parser->_i_st) {
+
+	__FSM_STATE(Req_Uri) {
```

Comment:
I do understand the original intent, but I also do think that it's bad practice to reuse the top-level states as interior states. As the code evolves, this may lead to very difficult bugs if, say, the values for top-level states are changed in a certain way. I am not saying that it will happen, perhaps not as this is still a small code base, but it's best to exclude the very possibility of a bug like this.

I believe it's best to introduce **interior** states for this function.

Reply:
@keshonok Thanks for review.
This code has been redesigned according to https://github.com/tempesta-tech/tempesta/pull/891#discussion_r165703517